### PR TITLE
add semaphore and illegal state exception to chat

### PR DESCRIPTION
**Pull request by davidmotson**

### CODE REVIEW

**Comment by daymxn** [comment]
don't forget to create a changelog

File: generativeai/src/main/java/com/google/ai/client/generativeai/Chat.kt [hunk 1]
```diff
@@ -39,6 +40,7 @@ import kotlinx.coroutines.flow.onEach
  * @property history the previous interactions with the model
  */
 class Chat(private val model: GenerativeModel, val history: MutableList<Content> = ArrayList()) {
+  private var lock = Semaphore(1)
 
   /**
    * Generates a response from the backend with the provided [Content], and any previous ones
```
Comment (rlazo) [inline generativeai/src/main/java/com/google/ai/client/generativeai/Chat.kt:46]: Update the refdoc to reflect the behaviour change.

File: generativeai/src/main/java/com/google/ai/client/generativeai/Chat.kt [hunk 2]
```diff
@@ -87,6 +91,7 @@ class Chat(private val model: GenerativeModel, val history: MutableList<Content>
    */
```
Comment (rlazo) [inline generativeai/src/main/java/com/google/ai/client/generativeai/Chat.kt:91]: Same here

File: generativeai/src/main/java/com/google/ai/client/generativeai/Chat.kt [hunk 3]
```diff
@@ -156,4 +162,13 @@ class Chat(private val model: GenerativeModel, val history: MutableList<Content>
       throw InvalidStateException("Chat prompts should come from the 'user' role.")
     }
   }
+
+  private fun attemptLock() {
+    if (!lock.tryAcquire()) {
+      throw IllegalStateException(
```
Comment (daymxn) [inline generativeai/src/main/java/com/google/ai/client/generativeai/Chat.kt:168]: Should be using our own variant, which is a subclass of `GoogleGenerativeAIException`

```suggestion
      throw InvalidStateException(
```

File: generativeai/src/main/java/com/google/ai/client/generativeai/Chat.kt [hunk 4]
```diff
@@ -36,6 +36,9 @@ import kotlinx.coroutines.flow.onEach
  * Handles the capturing and storage of the communication with the model, providing methods for
  * further interaction.
  *
+ * Note: This object is not thread-safe, and calling [sendMessage] multiple times without waiting
+ * for a response will throw an [InvalidStateException]
```
Comment (rlazo) [inline generativeai/src/main/java/com/google/ai/client/generativeai/Chat.kt:40]:
```suggestion
 * for a response will throw an [InvalidStateException].
```

**Review by rlazo — APPROVED** [review]

**Pull request merged**